### PR TITLE
Feature/random mutation operator

### LINKED ISSUE #5: Random Mutation Operator
Implement the random mutation operator for real-value genes. The operator simply adds a random number from a predefined interval to a gene. A newly introduced mutation_probability decides wether a single gene is mutated or not.
Idea: Implement the operator as callable type to define the mutation_probability and the interval.

### PR DESCRIPTION
**Pull request by domin7k**
Implements a callable struct containing a mutation rate and an interval of witch values are added with the probability of the mutation rate to an input matrix.
Currently supports Float64, Integer and Bool

### CODE REVIEW

File: src/Mutation.jl [hunk 1]
```diff
@@ -1,7 +1,58 @@
...
+    end
+
+end
+
+"""
+    Mutates the genes with a probability of c.mutation_rate and values in the interval c.mutation_interval
+"""
+function (c::RealGeneMutation)(genes::Matrix{T})::Matrix{T} where T<:Float64
+    
+    mask = rand((0,1), size(genes)) .< c.mutation_rate
```
Comment (Daniel1402) [inline src/Mutation.jl]: Use also Uniform(0, 1) to create the mask. `rand((0,1), size(genes))` does only produce 0s and 1s.

File: src/Mutation.jl [hunk 2]
```diff
@@ -1,7 +1,58 @@
...
+    random_additions = rand(Uniform(c.mutation_interval[1], c.mutation_interval[2]), size(genes))
+    return genes .+ (mask .&& random_additions)
+end
+
+"""
+    Mutates the genes with a probability of c.mutation_rate and values in the interval c.mutation_interval
+"""
+function (c::RealGeneMutation)(genes::Matrix{T})::Matrix{T} where T<:Integer
+    mask = rand((0,1), size(genes)) .< c.mutation_rate
```
Comment (Daniel1402) [inline src/Mutation.jl]: Use also Uniform(0, 1) to create the mask. `rand((0,1), size(genes))` does only produce 0s and 1s.

File: src/Mutation.jl [hunk 3]
```diff
@@ -1,7 +1,58 @@
...
+    random_additions = rand(range(c.mutation_interval[1], c.mutation_interval[2]), size(genes))
+    return genes .+ (mask .&& random_additions)
+end
+
+"""
+    Mutates the genes with a probability of c.mutation_rate and values in the interval c.mutation_interval
+"""
+function (c::RealGeneMutation)(genes::Matrix{T})::Matrix{T} where T<:Bool
+    mask = rand((0,1), size(genes)) .< c.mutation_rate
```
Comment (Daniel1402) [inline src/Mutation.jl]: Use also Uniform(0, 1) to create the mask. `rand((0,1), size(genes))` does only produce 0s and 1s.

File: test/test_mutation.jl [hunk 4]
```diff
@@ -1,4 +1,30 @@
...
+    for gene in mutated_genes
+        @test -0.5 <= gene <= 0.5
+    end
+    @test mutation.mutation_rate == 0.1
+    @test mutation.mutation_interval == (-0.5, 0.5)
+
+    mutation_int = RealGeneMutation(0.9, (0,17))
+    mutated_genes = mutation_int([0 0])
+    @test all(mutated_genes .>= 0 .& mutated_genes .<= 17)
+
```
Comment (Daniel1402) [inline test/test_mutation.jl:19]: Maybe orient your tests on `rouletteWheelSelection` where a rand_generator is passed to the function. This allows you wirte more meaningful tests

File: src/Mutation.jl [hunk 5]
```diff
@@ -1,7 +1,58 @@
 module Mutation
 
 using ..Types
+using Distributions
 
-# TODO
+"""
+    Defines the mutation of genes of real values (currently only Float64 and Integer types (including Bool))
+    Mutation is applied for each gene on its own with probability `mutation_rate` vith values from the uniform
```
Comment (Daniel1402) [inline src/Mutation.jl]: Typo `vith`

File: src/Mutation.jl [hunk 6]
```diff
@@ -1,7 +1,58 @@
...
+    distribution in the interval 'mutation_interval'.
```
Comment (Daniel1402) [inline src/Mutation.jl]: use \` istead of \'

File: src/Mutation.jl [hunk 7]
```diff
@@ -1,7 +1,58 @@
...
+        if mutation_interval[1] == mutation_interval[2]
+            throw(ArgumentError("Mutation interval must have a range"))
+        end
+        new{eltype(mutation_interval)}(mutation_rate, mutation_interval)
+    end
+
+end
+
+"""
+    Mutates the genes with a probability of c.mutation_rate and values in the interval c.mutation_interval
```
Comment (Daniel1402) [inline src/Mutation.jl]: add \` to c.mutation_interval

**Review by Daniel1402 — CHANGES_REQUESTED** [review]
Don't forget to pull, since I rebased your code.

File: src/Mutation.jl [hunk 8]
```diff
@@ -1,7 +1,58 @@
...
+"""
+    Defines the mutation of genes of real values (currently only Float64 and Integer types (including Bool))
+    Mutation is applied for each gene on its own with probability `mutation_rate` vith values from the uniform
+    distribution in the interval 'mutation_interval'.
+"""
+struct RealGeneMutation{T<:Union{Float64, Integer}}
+    mutation_rate::Float64
+    mutation_interval::Tuple{T, T}
 
+    function RealGeneMutation(mutation_rate::Float64, mutation_interval::Tuple{T, T}) where T<:Real
```
Comment (Daniel1402) [inline src/Mutation.jl]: Why do you have lower type restrictions on the constructor (Real) than on the struct itself  (only Float64 and Integer)?

File: src/Mutation.jl (hunk 2 above)
Comment (domin7k) [reply src/Mutation.jl:44]: s.o.

File: src/Mutation.jl (hunk 3 above)
Comment (domin7k) [reply src/Mutation.jl:53]: s.o.

File: src/Mutation.jl (hunk 6 above)
Comment (domin7k) [reply src/Mutation.jl:9]: ?

**Review by Daniel1402 — APPROVED** [review]

**Pull request merged**